samples/bot-calling-meeting/csharp/Source/CallingBotSample/Bots/CallingBot.cs

```diff
@@ -94,25 +103,86 @@ private void NotificationProcessor_OnNotificationReceived(NotificationEventArgs
+
             if (args.ResourceData is Call call)
             {
+                // If the notification is a newly created, incoming call, answer it
                 if (args.ChangeType == ChangeType.Created && call.State == CallState.Incoming)
                 {
-                    await callService.Answer(GetCallIdFromNotification(args), audioRecordingConstants.Speech, audioRecordingConstants.PleaseRecordYourMessage);
+                    await callService.Answer(callId, audioRecordingConstants.Speech, audioRecordingConstants.PleaseRecordYourMessage);
                 }
+                // If the notification is established (answered), fire a recording prompt
                 else if (
                     args.ChangeType == ChangeType.Updated
-                    && call.State == CallState.Established
-                    // The below helps to distinguish between two similar Established notifications that are sent for incoming 1:1 meetings
-                    && (call.MediaState?.Audio == MediaState.Active
-                    // And this covers calls that the bot creates
-                    || call.Direction == CallDirection.Outgoing))
+                    && call.State == CallState.Established)
+                {
+                    // Some scenarios fire two CallState.Established events. The use of a cache ensures we only play the prompt once on meeting join
+                    // This works by keeping track of when the record prompt on meeting established is sent, and if another notifications comes in
+                    // we do not send the prompt any more.
+                    string key = $"{callId}:established";
+                    if (!callBotCache.Get<bool>(key))
+                    {
+                        callBotCache.Set(key, true, new MemoryCacheEntryOptions
+                        {
+                            AbsoluteExpirationRelativeToNow = TimeSpan.FromSeconds(30)
+                        });
+                        await callService.Record(callId, audioRecordingConstants.PleaseRecordYourMessage);
+                    }
+                }
+            }
+            // If the notification is a recording, download from the Teams Recording Service, and then echo the audio back to the call
+            else if (args.ResourceData is RecordOperation recording)
+            {
+                if (recording.ResultInfo.Code >= 400)
                 {
-                    await callService.PlayPrompt(GetCallIdFromNotification(args), audioRecordingConstants.Speech);
+                    return;
+                }
+
+                var recordingLocation = await teamsRecordingService.DownloadRecording(recording.RecordingLocation, recording.RecordingAccessToken);
+
+                await callService.PlayPrompt(
+                    callId,
+                    new MediaInfo
+                    {
+                        Uri = new Uri(botOptions.BotBaseUrl, recordingLocation).ToString(),
```

Comment:
Is this required to be a public URL so it can be played back?

Reply:
Yes the Teams service will download the wav file so they need to be able to access it

Reply:
It's worth calling out the risk here, and also that in a production system you wouldn't want to download the file to the server's content directory. The randomized URL provides some defense, but make sure the directory is not set to allow enumeration.

I *think* a better solution might be to upload to a blob store, and use a SAS token to grant time-limited access to the recording. We don't have to implement that here, but we should specifically call this out.

Reply:
Yes that's a fair point, I've updated the comment here and expanded on it in a comment in the TeamsRecordingService